File: javascript/src/schema.ts
/** 4-way behavioural routing schema (version 1.0). */

export const SCHEMA_VERSION = "1.0";

export const PRIMARY_ACTIONS = ["answer_small", "rag", "tools", "escalate_large"] as const;
export type PrimaryAction = (typeof PRIMARY_ACTIONS)[number];

export const REASONING_LEVELS = ["low", "medium", "high"] as const;
export type ReasoningLevel = (typeof REASONING_LEVELS)[number];

export const DOMAINS = [
  "general_knowledge",
  "company_current",
  "tool_action",
  "multi_hop_planning",
  "ambiguous",
  "safety_mild",
  "open_domain_qa",
] as const;
export type Domain = (typeof DOMAINS)[number];

export const REQUIRED_FIELDS = [
  "id",
  "query",
  "primary_action",
  "needs_rag",
  "needs_tools",
  "reasoning_level",
  "rationale",
  "domain",
] as const;

export type RoutingRecord = {
  id: string;
  query: string;
  primary_action: string;
  needs_rag: boolean;
  needs_tools: boolean;
  reasoning_level: string;
  rationale: string;
  domain: string;
};

export class SchemaError extends Error {
  readonly errors: string[];

  constructor(errors: string[]) {
    const list = [...errors];
    super(list.length ? list.join("; ") : "invalid routing example");
    this.name = "SchemaError";
    this.errors = list;
  }
}

export function isPrimaryAction(value: unknown): value is PrimaryAction {
  return typeof value === "string" && (PRIMARY_ACTIONS as readonly string[]).includes(value);
}

export function parsePrimaryAction(value: string): PrimaryAction {
  if (!isPrimaryAction(value)) {
    throw new SchemaError([`primary_action '${value}' not in ${PRIMARY_ACTIONS}`]);
  }
  return value;
}

export class RoutingExample {
  readonly id: string;
  readonly query: string;
  readonly primary_action: string;
  readonly needs_rag: boolean;
  readonly needs_tools: boolean;
  readonly reasoning_level: string;
  readonly rationale: string;
  readonly domain: string;

  constructor(fields: RoutingRecord) {
    this.id = fields.id;
    this.query = fields.query;
    this.primary_action = fields.primary_action;
    this.needs_rag = fields.needs_rag;
    this.needs_tools = fields.needs_tools;
    this.reasoning_level = fields.reasoning_level;
    this.rationale = fields.rationale;
    this.domain = fields.domain;
  }

  toDict(): RoutingRecord {
    return {
      id: this.id,
      query: this.query,
      primary_action: this.primary_action,
      needs_rag: this.needs_rag,
      needs_tools: this.needs_tools,
      reasoning_level: this.reasoning_level,
      rationale: this.rationale,
      domain: this.domain,
    };
  }

  static fromDict(
    row: Record<string, unknown>,
    options: { validate?: boolean } = {},
  ): RoutingExample {
    const validate = options.validate ?? true;
    if (validate) {
      const errors = validateExample(row);
      if (errors.length) {
        throw new SchemaError(errors);
      }
    }
    return new RoutingExample({
      id: String(row.id),
      query: String(row.query),
      primary_action: String(row.primary_action),
      needs_rag: Boolean(row.needs_rag),
      needs_tools: Boolean(row.needs_tools),
      reasoning_level: String(row.reasoning_level),
      rationale: String(row.rationale),
      domain: String(row.domain),
    });
  }
}

export function validateExample(
  row: Record<string, unknown>,
  options: { lineNo?: number } = {},
): string[] {
  const prefix = options.lineNo != null ? `line ${options.lineNo}: ` : "";
  const errors: string[] = [];

  for (const field of REQUIRED_FIELDS) {
    if (!(field in row)) {
      errors.push(`${prefix}missing required field '${field}'`);
    }
  }
  if (errors.length) {
    return errors;
  }

  if (typeof row.id !== "string" || !row.id.trim()) {
    errors.push(`${prefix}id must be a non-empty string`);
  }
  if (typeof row.query !== "string" || !row.query.trim()) {
    errors.push(`${prefix}query must be a non-empty string`);
  }
  if (typeof row.rationale !== "string" || !row.rationale.trim()) {
    errors.push(`${prefix}rationale must be a non-empty string`);
  }

  const action = row.primary_action;
  if (!(PRIMARY_ACTIONS as readonly unknown[]).includes(action)) {
    errors.push(`${prefix}primary_action '${String(action)}' not in ${PRIMARY_ACTIONS}`);
  }

  if (typeof row.needs_rag !== "boolean") {
    errors.push(`${prefix}needs_rag must be bool, got ${typeName(row.needs_rag)}`);
  }
  if (typeof row.needs_tools !== "boolean") {
    errors.push(`${prefix}needs_tools must be bool, got ${typeName(row.needs_tools)}`);
  }

  const level = row.reasoning_level;
  if (!(REASONING_LEVELS as readonly unknown[]).includes(level)) {
    errors.push(`${prefix}reasoning_level '${String(level)}' not in ${REASONING_LEVELS}`);
  }

  const domain = row.domain;
  if (!(DOMAINS as readonly unknown[]).includes(domain)) {
    errors.push(`${prefix}domain '${String(domain)}' not in ${DOMAINS}`);
  }

  if (action === "rag" && row.needs_rag !== true) {
    errors.push(`${prefix}primary_action=rag requires needs_rag=true`);
  }
  if (action === "tools" && row.needs_tools !== true) {
    errors.push(`${prefix}primary_action=tools requires needs_tools=true`);
  }

  return errors;
}

function typeName(value: unknown): string {
  if (value === null) {
    return "null";
  }
  return typeof value === "object" ? value.constructor?.name || "object" : typeof value;
}

File: src/fta_router/schema.py
"""4-way behavioural routing schema (version 1.0)."""

from __future__ import annotations

from collections.abc import Mapping
from dataclasses import asdict, dataclass
from typing import Any, Final

SCHEMA_VERSION: Final = "1.0"

PRIMARY_ACTIONS: Final[tuple[str, ...]] = (
    "answer_small",
    "rag",
    "tools",
    "escalate_large",
)
REASONING_LEVELS: Final[tuple[str, ...]] = ("low", "medium", "high")
DOMAINS: Final[tuple[str, ...]] = (
    "general_knowledge",
    "company_current",
    "tool_action",
    "multi_hop_planning",
    "ambiguous",
    "safety_mild",
    "open_domain_qa",
)

REQUIRED_FIELDS: Final[tuple[str, ...]] = (
    "id",
    "query",
    "primary_action",
    "needs_rag",
    "needs_tools",
    "reasoning_level",
    "rationale",
    "domain",
)


class SchemaError(ValueError):
    """Raised when a routing record fails schema validation."""

    def __init__(self, errors: list[str]) -> None:
        self.errors = list(errors)
        super().__init__("; ".join(self.errors) if self.errors else "invalid routing example")


def is_primary_action(value: object) -> bool:
    """Return True if *value* is one of the four routing actions."""
    return value in PRIMARY_ACTIONS


def parse_primary_action(value: str) -> str:
    """Return *value* if it is a valid primary action, else raise ``SchemaError``."""
    if not is_primary_action(value):
        raise SchemaError([f"primary_action '{value}' not in {PRIMARY_ACTIONS}"])
    return value


@dataclass(frozen=True)
class RoutingExample:
    """One labelled behavioural routing decision."""

    id: str
    query: str
    primary_action: str
    needs_rag: bool
    needs_tools: bool
    reasoning_level: str
    rationale: str
    domain: str

    def to_dict(self) -> dict[str, Any]:
        return asdict(self)

    @classmethod
    def from_dict(cls, row: Mapping[str, Any], *, validate: bool = True) -> RoutingExample:
        if validate:
            errors = validate_example(row)
            if errors:
                raise SchemaError(errors)
        return cls(
            id=str(row["id"]),
            query=str(row["query"]),
            primary_action=str(row["primary_action"]),
            needs_rag=bool(row["needs_rag"]),
            needs_tools=bool(row["needs_tools"]),
            reasoning_level=str(row["reasoning_level"]),
            rationale=str(row["rationale"]),
            domain=str(row["domain"]),
        )


def validate_example(row: Mapping[str, Any], *, line_no: int | None = None) -> list[str]:
    """Return a list of validation error strings (empty if valid)."""
    prefix = f"line {line_no}: " if line_no is not None else ""
    errors: list[str] = []

    for field in REQUIRED_FIELDS:
        if field not in row:
            errors.append(f"{prefix}missing required field '{field}'")
    if errors:
        return errors

    if not isinstance(row["id"], str) or not row["id"].strip():
        errors.append(f"{prefix}id must be a non-empty string")
    if not isinstance(row["query"], str) or not row["query"].strip():
        errors.append(f"{prefix}query must be a non-empty string")
    if not isinstance(row["rationale"], str) or not row["rationale"].strip():
        errors.append(f"{prefix}rationale must be a non-empty string")

    action = row["primary_action"]
    if action not in PRIMARY_ACTIONS:
        errors.append(f"{prefix}primary_action '{action}' not in {PRIMARY_ACTIONS}")

    if not isinstance(row["needs_rag"], bool):
        errors.append(f"{prefix}needs_rag must be bool, got {type(row['needs_rag']).__name__}")
    if not isinstance(row["needs_tools"], bool):
        errors.append(f"{prefix}needs_tools must be bool, got {type(row['needs_tools']).__name__}")

    level = row["reasoning_level"]
    if level not in REASONING_LEVELS:
        errors.append(f"{prefix}reasoning_level '{level}' not in {REASONING_LEVELS}")

    domain = row["domain"]
    if domain not in DOMAINS:
        errors.append(f"{prefix}domain '{domain}' not in {DOMAINS}")

    if action == "rag" and row.get("needs_rag") is not True:
        errors.append(f"{prefix}primary_action=rag requires needs_rag=true")
    if action == "tools" and row.get("needs_tools") is not True:
        errors.append(f"{prefix}primary_action=tools requires needs_tools=true")

    return errors
